File: pkg/trajectory_ir/runtime/sandbox.py
"""Sandbox / what-if trajectory mode (README §10 R06).

Live mode is the default. Sandbox mode allows planning and safe tools
(PURE, READ_ONLY, IDEMPOTENT_WRITE) but rejects NON_IDEMPOTENT_WRITE,
AGENT_SPAWN, and SENSITIVE effects before the tool body runs.
"""

from __future__ import annotations

from enum import StrEnum

from trajectory_ir.effects import EffectClass, is_forbidden_in_sandbox


class RunMode(StrEnum):
    LIVE = "live"
    SANDBOX = "sandbox"


class SandboxForbidden(Exception):
    """Raised when sandbox mode rejects a tool effect class (R06)."""

    def __init__(self, tool_name: str, effect_class: EffectClass):
        self.tool_name = tool_name
        self.effect_class = effect_class
        super().__init__(
            f"SANDBOX_FORBIDDEN: tool {tool_name!r} "
            f"effect={effect_class.value} is forbidden in sandbox mode"
        )


def normalize_run_mode(mode: RunMode | str | None) -> RunMode:
    if mode is None:
        return RunMode.LIVE
    if isinstance(mode, RunMode):
        return mode
    m = str(mode).strip().lower()
    if m in ("live", ""):
        return RunMode.LIVE
    if m == "sandbox":
        return RunMode.SANDBOX
    raise ValueError(f"unsupported run mode {mode!r}; use live or sandbox")


def assert_tool_allowed_in_mode(
    mode: RunMode | str | None,
    *,
    tool_name: str,
    effect_class: EffectClass,
) -> None:
    """Raise SandboxForbidden if this tool must not run under the given mode."""
    run_mode = normalize_run_mode(mode)
    if run_mode is RunMode.SANDBOX and is_forbidden_in_sandbox(effect_class):
        raise SandboxForbidden(tool_name, effect_class)

File: go/trajir/sandbox/sandbox.go
// Package sandbox implements live vs sandbox run modes (R06).
package sandbox

import (
	"fmt"
	"strings"

	"github.com/Coder-s-OG-s/Trajectory-IR/go/trajir/effects"
)

// Mode is the trajectory execution mode.
type Mode string

const (
	ModeLive    Mode = "live"
	ModeSandbox Mode = "sandbox"
)

// Forbidden is raised when sandbox rejects a tool effect.
type Forbidden struct {
	ToolName string
	Effect   effects.EffectClass
}

func (e *Forbidden) Error() string {
	return fmt.Sprintf(
		"SANDBOX_REJECTS_NON_IDEMPOTENT_WRITE: tool %q effect=%s is forbidden in sandbox mode",
		e.ToolName,
		e.Effect,
	)
}

// NormalizeMode maps string/empty to ModeLive or ModeSandbox.
func NormalizeMode(mode string) (Mode, error) {
	m := strings.TrimSpace(strings.ToLower(mode))
	if m == "" || m == "live" {
		return ModeLive, nil
	}
	if m == "sandbox" {
		return ModeSandbox, nil
	}
	return "", fmt.Errorf("unsupported run mode %q; use live or sandbox", mode)
}

// AssertToolAllowed raises Forbidden in sandbox for gated effects.
func AssertToolAllowed(mode Mode, toolName string, effect effects.EffectClass) error {
	if mode == ModeSandbox && effects.RequiresBlockAndGate(effect) {
		return &Forbidden{ToolName: toolName, Effect: effect}
	}
	return nil
}
